File: udemy/ds-and-a/csharp-solutions/DSAndA/StacksAndQueues/Assignment3/QueueUsingLinkedListAsUnderlyingDS.cs
using DSAndA.LinkedLists;

namespace DSAndA.StacksAndQueues.Assignment3
{
    public class QueueUsingLinkedListAsUnderlyingDS
    {
        private Node<int> head;
        private Node<int> tail;

        private bool IsEmpty => this.tail == null;

        public void Push(int element)
        {
            Node<int> newNode = new Node<int>(element);

            if (!this.IsEmpty)
                this.tail.NextNode = newNode;
            else
            {
                this.head = newNode;
                this.tail = newNode;
            }
        }

        public int Pop()
        {
            int poppedElement = this.head.GetData();

            this.head = this.head.NextNode;

            return poppedElement;
        }

        public int Peek()
        {
            return this.head.GetData();
        }
    }
}

File: udemy/ds-and-a/javascript-solutions/src/StackAndQueues/Assignment3/QueueUsingLinkedListAsUnderlyingDS.js
const Node = require('../../LinkedLists/Node');

class QueueUsingLinkedListAsUnderlyingDS {
    constructor() {
        this.head = null;
        this.tail = null;
    }

    enqueue(data) {
        let newNode = new Node(data);

        if (!this.head) {
            this.head = newNode;
            this.tail = newNode;
        } else {
            this.tail.nextNode = newNode;
            this.tail = newNode;
        }
    }

    dequeue() {
        this.head = this.head.nextNode;
    }

    toString() {
        let result = [];
        let current = this.head;

        while (current != null) {
            result.push(current.data);

            current = current.nextNode;
        }

        return result.join(', ');
    }
}

module.exports = QueueUsingLinkedListAsUnderlyingDS;